### DISCUSSION TITLE
[v8]: State updater type

### DISCUSSION THREAD

**Discussion by Moshyfawn** [TanStack/table#3899] [Q&A]
I'm figuring out a controlled table config _(RT + RL, URL state)_: I'm getting a TS error trying to integrate the state updates via the state change handlers.
```zsh
This expression is not callable.
  Not all constituents of type 'Updater<PaginationState>' are callable.
    Type 'PaginationState' has no call signatures.ts(2349)
```

The idea is to update the URL `onPaginationChange`
```js
const { pagination } = useSearch()
const { pageIndex = 0, pageSize = 10 } = pagination || {}
const navigate = useNavigate()

const instance = useTableInstance(table, {
  state: {
    pagination: {
      pageIndex,
      pageSize,
      pageCount
    }
  },
  onPaginationChange: (setPagination) => {
    navigate({
      search: (currentSearch) => ({
        ...currentSearch,
        // INFO: the TS error occures here
        pagination: setPagination({
          pageIndex,
          pageSize,
          pageCount
        })
      })
    })
  }
})
```

Any idea on how to correctly go about it?

Here's [a codesanbox example](https://codesandbox.io/s/react-table-url-state-1gzdzl?file=/src/App.tsx)_(see the `onPaginationChange` callback)_

_P.S. I'm currently trying the v8 RT version. It's insanely convenient and I'm loving the UX!_

**Comment by KevinVandy** [comment]
Try this
```ts
onPaginationChange: (updater) => setPagination((old) => functionalUpdate(updater, old)),
```

**Reply by GuidovdRiet** [reply]
Hi! Thank you. Can this be added to the documentation? Hard one to understand what exactly is going on here.

**Comment by nicolasrouanne** [comment]
I find the signature of the `onPaginationChange` callback hard to grasp when not using a classic `useState` setter. In my case I'm saving the pagination state in a _Context_ (to be able to use it in deeply nested components).

Here is how I got it working:
```tsx
const [filters, setFilters] = useContext(TableContext);
const { pageIndex, pageSize } = filters;

<DataTable
  onPaginationChange={(updater) =>
    setFilters({
      ...filters,
      ...updater({ pageIndex, pageSize }),
    })
  }
  state={{ pagination: { pageIndex, pageSize } }}
/>;
```

**Reply by nicolasrouanne** [reply]
I still have the TS error from the original post though. And I can't wrap my head around it 🙈

```
This expression is not callable.
  Not all constituents of type 'Updater<PaginationState>' are callable.
    Type 'PaginationState' has no call signatures.ts(2349)
```
It's because `Updater` can be either a value or a function
```ts
export type Updater<T> = T | ((old: T) => T);
```

I don't really know how to adapt [the following theoretical](https://github.com/TanStack/table/discussions/3899#discussioncomment-2883855) definition for my usecase:
```ts
onPaginationChange: (updater) => setPagination((old) => functionalUpdate(updater, old)),
```

This (inspired by [this StackOverflow comment](https://stackoverflow.com/questions/74721400/tanstack-table-how-can-i-use-redux-action-in-onpaginationchange-instead-of-set#comment134935083_74977208)) fixes it, but feels like a hack:
```ts
onPaginationChange={(updater) => {
  if (typeof updater === "function")
    setFilters({
      ...filters,
      ...updater({ pageIndex, pageSize }),
    });
}}
```
